src/Data/Bin/Base.agda

```diff
@@ -0,0 +1,109 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Arithmetic of binary represented natural numbers. Initial part.
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Data.Bin.Base where
+
+open import Algebra.FunctionProperties using (Op₂)
+open import Data.Nat        using (ℕ) renaming (suc to 1+_; _*_ to _*ₙ_)
+open import Function        using (_∘_)
+open import Relation.Binary using (Decidable)
+open import Relation.Binary.PropositionalEquality using (_≡_; refl; cong)
+open import Relation.Nullary                      using (yes; no)
+
+
+data Bin : Set
+  where
+  0#    : Bin
+  2suc  : Bin → Bin    -- \n→ 2*(1+n)  arbitrary nonzero even
+  suc2* : Bin → Bin    -- \n→ 1 + 2n   arbitrary odd
+--
+-- This representation is unique for each natural number.
+
+
+size : Bin → ℕ       -- (number of constructors) - 1.
+                       -- It is used in some termination proofs.
+size 0#        = 0
+size (2suc x)  = 1+ (size x)
+size (suc2* x) = 1+ (size x)
+
+
+------------------------------------------------------------------------------
+-- Arithmetic operations on Bin.
+------------------------------------------------------------------------------
+
+suc : Bin → Bin
+suc 0#        =  suc2* 0#
+suc (2suc x)  =  suc2* (suc x)   -- 1 + 2(1+x)
+suc (suc2* x) =  2suc x          -- 1 + 1 + 2x =  2*(1+x)
+
+fromℕ : ℕ → Bin                  -- it costs O(n)
+fromℕ 0      = 0#
+fromℕ (1+ n) = suc (fromℕ n)
+
+
+------------------------------------------------------------------------------
+-- Constants
+------------------------------------------------------------------------------
+
+1B = suc 0#
+2B = suc 1B
+3B = suc 2B
+4B = suc 3B
+5B = suc 4B
+6B = suc 5B
+7B = suc 6B
+8B = suc 7B
+9B = suc 8B
+
```

Comment:
Can you please make all of these `pattern`s rather than mere definitions?
This way they can be used in the left hand side to. They are a bit more
annoying to define but more convenient to use.

i.e.

```agda
pattern 1B = suc2* 0B
pattern 2B = 2*suc 0B
pattern 3B = suc2* 1B
pattern 4B = 2*suc 1B
pattern 5B = suc2* 2B
pattern 6B = 2*suc 2B
pattern 7B = suc2* 3B
pattern 8B = 2*suc 3B
pattern 9B = suc2* 4B
```

Reply:
Given #830 I've decided to leave this one to the future.